### PR TITLE
Operators reloading info

### PR DESCRIPTION
**Pull request by Asleeepp**
### Description

Adds a config option to allow all players with the permission, to see whenever a player reloads a script, and the info that comes along with it, such as errors and/or warnings, etc.

---
**Target Minecraft Versions:** any 
**Requirements:** none 
**Related Issues:** https://github.com/SkriptLang/Skript/issues/6736

### CODE REVIEW

File: src/main/java/ch/njol/skript/SkriptCommand.java [hunk 1]
```diff
@@ -90,32 +94,48 @@ public class SkriptCommand implements CommandExecutor {
...
-			Skript.info(sender, StringUtils.fixCapitalization(PluralizingArgsMessage.format(m_reloaded.toString(what, timeTaken))));
-		else
-			Skript.error(sender, StringUtils.fixCapitalization(PluralizingArgsMessage.format(m_reload_error.toString(what, r.numErrors(), timeTaken))));
+
+	private static void reloaded(CommandSender sender, RedirectingLogHandler logHandler, TimingLogHandler timingLogHandler, String what, Object... args) {
+		what = args.length == 0 ? Language.get(CONFIG_NODE + ".reload." + what) : format(Language.format(CONFIG_NODE + ".reload." + what, args));
+		String timeTaken = String.valueOf(timingLogHandler.getTimeTaken());
+
+		String message;
+		String errorDetails = null;
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/SkriptCommand.java:118]: What's this doing?

File: src/main/java/ch/njol/skript/SkriptCommand.java [hunk 2]
```diff
@@ -90,32 +94,48 @@ public class SkriptCommand implements CommandExecutor {
...
+
+		if (SkriptConfig.sendReloadingInfoToOps.value())
+			notifyOperators(message, sender, logHandler);
 	}
-	
+
+
 	private static void info(CommandSender sender, String what, Object... args) {
-		what = args.length == 0 ? Language.get(CONFIG_NODE + "." + what) : PluralizingArgsMessage.format(Language.format(CONFIG_NODE + "." + what, args));
+		what = args.length == 0 ? Language.get(CONFIG_NODE + "." + what) : format(Language.format(CONFIG_NODE + "." + what, args));
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/SkriptCommand.java:133]: Why was this removed?

File: src/main/java/ch/njol/skript/SkriptCommand.java [hunk 3]
```diff
@@ -90,32 +94,48 @@ public class SkriptCommand implements CommandExecutor {
...
 		Skript.info(sender, StringUtils.fixCapitalization(what));
 	}
-	
+
 	private static void error(CommandSender sender, String what, Object... args) {
-		what = args.length == 0 ? Language.get(CONFIG_NODE + "." + what) : PluralizingArgsMessage.format(Language.format(CONFIG_NODE + "." + what, args));
+		what = args.length == 0 ? Language.get(CONFIG_NODE + "." + what) : format(Language.format(CONFIG_NODE + "." + what, args));
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/SkriptCommand.java:138]: likewise

File: src/main/java/ch/njol/skript/SkriptCommand.java [hunk 4]
```diff
@@ -537,5 +531,38 @@ private static Set<File> toggleFiles(File folder, boolean enable) throws IOExcep
 
 		return changed;
 	}
-	
+
+	private static void notifyOperators(String message, CommandSender sender, RedirectingLogHandler logHandler) {
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/SkriptCommand.java:535]: should add some javadocs

File: src/main/java/ch/njol/skript/SkriptCommand.java [hunk 5]
```diff
@@ -537,5 +531,38 @@ private static Set<File> toggleFiles(File folder, boolean enable) throws IOExcep
...
+						Skript.info(player, message);
+					} else {
+						for (String errorMsg : logHandler.getErrors()) {
+							SkriptLogger.sendFormatted(player, errorMsg);
+						}
+						Skript.error(player, message);
+					}
+				});
+
+			// Send the message to the console if the boolean is false
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/SkriptCommand.java:550]: this comment indicates that the brackets for the enclosing `if` are wrong

File: src/main/java/ch/njol/skript/SkriptCommand.java [hunk 6]
```diff
@@ -537,5 +531,38 @@ private static Set<File> toggleFiles(File folder, boolean enable) throws IOExcep
 
 		return changed;
 	}
-	
+
+	private static void notifyOperators(String message, CommandSender sender, RedirectingLogHandler logHandler) {
+		if (SkriptConfig.sendReloadingInfoToOps.value()) {
+			Bukkit.getOnlinePlayers().stream()
+				.filter(player -> player.isOp() && player.hasPermission("skript.reloadnotify") && !player.equals(sender))
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/SkriptCommand.java:538]: should this really check for isOp? Just checking for the permission would be more flexible for the user.

**Review by sovdeeth — CHANGES_REQUESTED** [review]
preliminary look-through

File: src/main/java/ch/njol/skript/SkriptCommand.java (hunk 5 above)
Comment (Asleeepp) [reply src/main/java/ch/njol/skript/SkriptCommand.java:550]: this comment i accidentally left, it was in an earlier point of this PR, will remove

File: src/main/java/ch/njol/skript/SkriptCommand.java [hunk 7]
```diff
@@ -54,8 +55,10 @@
 import java.util.Set;
 import java.util.stream.Collectors;
 
+import static ch.njol.skript.localization.PluralizingArgsMessage.format;
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/SkriptCommand.java:58]: don't import this statically
it was fine before, and now I think it looks like `format` is a method on this class instead.

**Review by sovdeeth — CHANGES_REQUESTED** [review]
mostly good but I would like to see an example of what it looks like.

**Review by sovdeeth — APPROVED** [review]

File: src/main/resources/plugin.yml [hunk 8]
```diff
@@ -53,6 +53,9 @@ permissions:
     skript.admin:
         default: op
         description: Allows to use administrative commands and to recieve notifications of new versions.
+    skript.reloadnotify:
+        default: op
```
Comment (Moderocky) [inline src/main/resources/plugin.yml:57]: I think this should default to false like effect commands do.

**Review by Moderocky — CHANGES_REQUESTED** [review]

File: src/main/resources/config.sk [hunk 9]
```diff
@@ -80,6 +80,10 @@ log effect commands: false
 # Whether Skript should log the usage of effect commands.
 # They will be logged as [INFORMATION] in this format: '<player> issued effect command: <command>'
 
+send reloading info to ops: false
+# Tells Skript whether to alert operators about when someone reloads a script file.
```
Comment (APickledWalrus) [inline src/main/resources/config.sk:84]:
```suggestion
# Whether Skript should alert operators when someone reloads a script file.
```

File: src/main/java/ch/njol/skript/SkriptCommand.java [hunk 10]
```diff
@@ -90,30 +92,45 @@ public class SkriptCommand implements CommandExecutor {
 		if (TestMode.DEV_MODE)
 			SKRIPT_COMMAND_HELP.add("test");
 	}
-	
-	private static void reloading(CommandSender sender, String what, Object... args) {
+
+	private static void reloading(CommandSender sender, String what, RedirectingLogHandler logHandler, Object... args) {
+		String prefix = args.length == 0 ? Language.get(CONFIG_NODE + ".reload." + "player reload") : Language.format(CONFIG_NODE + ".reload." + "player reload", sender.getName());
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/SkriptCommand.java:97]: This is only needed if sendReloadingInfoToOps is true, right?

File: src/main/java/ch/njol/skript/SkriptCommand.java [hunk 11]
```diff
@@ -90,30 +92,45 @@ public class SkriptCommand implements CommandExecutor {
...
 		what = args.length == 0 ? Language.get(CONFIG_NODE + ".reload." + what) : Language.format(CONFIG_NODE + ".reload." + what, args);
-		Skript.info(sender, StringUtils.fixCapitalization(m_reloading.toString(what)));
+		String message = StringUtils.fixCapitalization(m_reloading.toString(what));
+		Skript.info(sender, message);
+		if (SkriptConfig.sendReloadingInfoToOps.value()) {
+			String lowerCaseMessage = lowerCaseFirstChar(message);
+			notifyOperators(prefix + " " + lowerCaseMessage, sender, logHandler);
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/SkriptCommand.java:103]: I don't think it is a good idea to combine two language entries together like this. While this works for English, there are no guarantees for other languages. It'd be better to have the lang entry represent the entire message.

File: src/main/java/ch/njol/skript/log/RedirectingLogHandler.java [hunk 12]
```diff
@@ -30,31 +31,35 @@
...
 	}
-	
+
 	public int numErrors() {
 		return numErrors;
 	}
 
+	public List<String> getErrors() {
+		return new ArrayList<>(errorMessages);
+	}
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/log/RedirectingLogHandler.java:64]:
```suggestion
	}

```

File: src/main/java/ch/njol/skript/log/RedirectingLogHandler.java [hunk 13]
```diff
@@ -30,31 +31,35 @@
 public class RedirectingLogHandler extends LogHandler {
 
 	private final CommandSender recipient;
-	
 	private final String prefix;
-	
 	private int numErrors = 0;
-	
+	private final List<String> errorMessages = new ArrayList<>();
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/log/RedirectingLogHandler.java:36]: You should override `onStop` to clear this list. I do not think we want its contents to remain when the loghandler is stopped.

**Review by APickledWalrus — CHANGES_REQUESTED** [review]
I'm thinking in general it might be better to instead support multiple command senders in RedirectingLogHandler. It would avoid some of the more awkward changes.

File: src/main/java/ch/njol/skript/log/RedirectingLogHandler.java [hunk 14]
```diff
@@ -18,41 +18,41 @@
...
+		this.recipients = recipients;
 		this.prefix = prefix == null ? "" : prefix;
 	}
-	
+
 	@Override
 	public LogResult log(LogEntry entry) {
-		SkriptLogger.sendFormatted(recipient, prefix + entry.toFormattedString());
-		if (entry.level == Level.SEVERE)
-			numErrors++;
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/log/RedirectingLogHandler.java:47]: You should keep this

File: src/main/java/ch/njol/skript/log/RedirectingLogHandler.java [hunk 15]
```diff
@@ -18,41 +18,41 @@
...
 import org.bukkit.command.CommandSender;
 import org.eclipse.jdt.annotation.Nullable;
 
+import java.util.ArrayList;
+import java.util.List;
 import java.util.logging.Level;
 
 /**
- * Redirects the log to a {@link CommandSender}.
+ * Redirects the log to multiple {@link CommandSender}s.
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/log/RedirectingLogHandler.java:29]:
```suggestion
 * Redirects the log to one or more {@link CommandSender}s.
```

File: src/main/java/ch/njol/skript/log/RedirectingLogHandler.java [hunk 16]
```diff
@@ -18,41 +18,41 @@
...
 public class RedirectingLogHandler extends LogHandler {
 
-	private final CommandSender recipient;
-	
-	private final String prefix;
-	
+	private final List<CommandSender> recipients;
 	private int numErrors = 0;
-	
-	public RedirectingLogHandler(CommandSender recipient, @Nullable String prefix) {
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/log/RedirectingLogHandler.java:38]: We should keep this constructor in case any addons are using it.

File: src/main/java/ch/njol/skript/SkriptCommand.java [hunk 17]
```diff
@@ -124,15 +147,25 @@ public boolean onCommand(CommandSender sender, Command command, String label, St
 		if (!SKRIPT_COMMAND_HELP.test(sender, args))
 			return true;
 
+		boolean isReload = label.equalsIgnoreCase("reload");
+		List<CommandSender> recipients = new ArrayList<>();
+
+		if (isReload) {
+			recipients.add(sender);
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/SkriptCommand.java:154]: sender always need to be included

File: src/main/java/ch/njol/skript/SkriptCommand.java [hunk 18]
```diff
@@ -124,15 +147,25 @@ public boolean onCommand(CommandSender sender, Command command, String label, St
 		if (!SKRIPT_COMMAND_HELP.test(sender, args))
 			return true;
 
+		boolean isReload = label.equalsIgnoreCase("reload");
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/SkriptCommand.java:150]: does this work? i believe you need to check `args[0].equalsIgnoreCase("reload")`

File: src/main/java/ch/njol/skript/SkriptCommand.java [hunk 19]
```diff
@@ -90,30 +95,48 @@ public class SkriptCommand implements CommandExecutor {
...
+		}
 
-		if (r.numErrors() == 0)
-			Skript.info(sender, StringUtils.fixCapitalization(PluralizingArgsMessage.format(m_reloaded.toString(what, timeTaken))));
-		else
-			Skript.error(sender, StringUtils.fixCapitalization(PluralizingArgsMessage.format(m_reload_error.toString(what, r.numErrors(), timeTaken))));
+		String playerReloadMesssage = Language.format(CONFIG_NODE + ".reload." + "player reload", sender.getName(), what);
+		// Notify operators
+		if (SkriptConfig.sendReloadingInfoToOps.value())
+			logHandler.log(new LogEntry( Level.INFO, playerReloadMesssage));
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/SkriptCommand.java:130]: you should only need to log in `reloading()`. This is when the reloading is finished

File: src/main/java/ch/njol/skript/SkriptCommand.java [hunk 20]
```diff
@@ -90,30 +95,48 @@ public class SkriptCommand implements CommandExecutor {
...
+
+	private static void reloading(CommandSender sender, String what, RedirectingLogHandler logHandler, Object... args) {
 		what = args.length == 0 ? Language.get(CONFIG_NODE + ".reload." + what) : Language.format(CONFIG_NODE + ".reload." + what, args);
-		Skript.info(sender, StringUtils.fixCapitalization(m_reloading.toString(what)));
+		String message = StringUtils.fixCapitalization(m_reloading.toString(what));
+		Skript.info(sender, message);
+
+		// Log reloading message
+		if (SkriptConfig.sendReloadingInfoToOps.value())
+			logHandler.log(new LogEntry(Level.INFO, Language.format(CONFIG_NODE + ".reload." + "player reload", sender.getName(), what)));
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/SkriptCommand.java:106]: does this still log? it would also send to the command sender. it might be better to send this message when building the recipients list.

**Review by APickledWalrus — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/log/RedirectingLogHandler.java [hunk 21]
```diff
@@ -18,43 +18,53 @@
...
 		this.prefix = prefix == null ? "" : prefix;
 	}
-	
+
 	@Override
 	public LogResult log(LogEntry entry) {
-		SkriptLogger.sendFormatted(recipient, prefix + entry.toFormattedString());
-		if (entry.level == Level.SEVERE)
+		String formattedMessage = prefix + entry.toFormattedString();
+		if (recipients != null) {
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/log/RedirectingLogHandler.java:50]: null check shouldn't be needed

File: src/main/java/ch/njol/skript/SkriptCommand.java [hunk 22]
```diff
@@ -124,15 +146,25 @@ public boolean onCommand(CommandSender sender, Command command, String label, St
 		if (!SKRIPT_COMMAND_HELP.test(sender, args))
 			return true;
 
+		boolean isReload = args[0].equalsIgnoreCase("reload");
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/SkriptCommand.java:149]: looks like this is one used once so you can just remove it and check directly instead

File: src/main/java/ch/njol/skript/SkriptCommand.java [hunk 23]
```diff
@@ -90,30 +95,47 @@ public class SkriptCommand implements CommandExecutor {
...
+
+	private static void reloading(CommandSender sender, String what, RedirectingLogHandler logHandler, Object... args) {
 		what = args.length == 0 ? Language.get(CONFIG_NODE + ".reload." + what) : Language.format(CONFIG_NODE + ".reload." + what, args);
-		Skript.info(sender, StringUtils.fixCapitalization(m_reloading.toString(what)));
+		String message = StringUtils.fixCapitalization(m_reloading.toString(what));
+		Skript.info(sender, message);
+
+		// Log reloading message
+		if (SkriptConfig.sendReloadingInfoToOps.value())
+			logHandler.log(new LogEntry(Level.INFO, Language.format(CONFIG_NODE + ".reload." + "player reload", sender.getName(), what)));
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/SkriptCommand.java:106]: this will also send to `sender`. i'm not sure if that's a big issue though.

File: src/main/java/ch/njol/skript/SkriptCommand.java [hunk 24]
```diff
@@ -90,30 +95,47 @@ public class SkriptCommand implements CommandExecutor {
...
+			Skript.error(sender, message);
+		}
 
-		if (r.numErrors() == 0)
-			Skript.info(sender, StringUtils.fixCapitalization(PluralizingArgsMessage.format(m_reloaded.toString(what, timeTaken))));
-		else
-			Skript.error(sender, StringUtils.fixCapitalization(PluralizingArgsMessage.format(m_reload_error.toString(what, r.numErrors(), timeTaken))));
+		// Notify operators
+		if (SkriptConfig.sendReloadingInfoToOps.value())
+			logHandler.log(new LogEntry( Level.INFO, message));
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/SkriptCommand.java:129]: This would send to the `sender` twice. It would be better to do something like:
```
if no errors:
  logHandler.log(message) // log with INFO
else:
  logHandler.log(message) // log with SEVERE
```
the log handler will already handle sending to operators so you don't need to check it again

**Review by APickledWalrus — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/sections/SecFilter.java
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/sections/SecFilter.java:1]: what's this guy

File: src/main/java/ch/njol/skript/SkriptCommand.java [hunk 25]
```diff
@@ -90,30 +95,43 @@ public class SkriptCommand implements CommandExecutor {
...
+			logHandler.log(new LogEntry(Level.INFO, message));
+		} else {
+			message = StringUtils.fixCapitalization(PluralizingArgsMessage.format(m_reload_error.toString(what, logHandler.numErrors(), timeTaken)));
+			logHandler.log(new LogEntry(Level.SEVERE, message));
+		}
 	}
-	
+
+
+
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/SkriptCommand.java:129]:
```suggestion

```
Extra whitespace can be removed

File: src/main/java/ch/njol/skript/SkriptCommand.java [hunk 26]
```diff
@@ -90,30 +95,43 @@ public class SkriptCommand implements CommandExecutor {
...
+		Skript.info(sender, message);
+
+		// Log reloading message
+		if (SkriptConfig.sendReloadingInfoToOps.value())
+			logHandler.log(new LogEntry(Level.INFO, Language.format(CONFIG_NODE + ".reload." + "player reload", sender.getName(), what)));
 	}
-	
+
+
+
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/SkriptCommand.java:110]:
```suggestion

```
Extra whitespace can be removed

File: src/main/java/ch/njol/skript/SkriptCommand.java [hunk 27]
```diff
@@ -124,15 +142,24 @@ public boolean onCommand(CommandSender sender, Command command, String label, St
 		if (!SKRIPT_COMMAND_HELP.test(sender, args))
 			return true;
 
+		List<CommandSender> recipients = new ArrayList<>();
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/SkriptCommand.java:145]: This could be a Set, which would mean you don't have to do the .equals check for each player.

**Review by APickledWalrus — APPROVED** [review]

File: src/main/java/ch/njol/skript/log/RedirectingLogHandler.java [hunk 28]
```diff
@@ -18,43 +18,53 @@
...
- * Redirects the log to a {@link CommandSender}.
+ * Redirects the log to one or more {@link CommandSender}s.
  */
 public class RedirectingLogHandler extends LogHandler {
 
-	private final CommandSender recipient;
-	
-	private final String prefix;
-	
+	private final List<CommandSender> recipients;
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/log/RedirectingLogHandler.java:36]:
```suggestion
	private final Collection<CommandSender> recipients;
```

File: src/main/java/ch/njol/skript/log/RedirectingLogHandler.java [hunk 29]
```diff
@@ -18,43 +18,53 @@
...
 	private int numErrors = 0;
-	
+	private final String prefix;
+
 	public RedirectingLogHandler(CommandSender recipient, @Nullable String prefix) {
-		this.recipient = recipient;
+		this(Collections.singletonList(recipient), prefix);
+	}
+
+	public RedirectingLogHandler(List<CommandSender> recipients, @Nullable String prefix) {
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/log/RedirectingLogHandler.java:44]:
```suggestion
	public RedirectingLogHandler(Collection<CommandSender> recipients, @Nullable String prefix) {
```

**Review by sovdeeth — CHANGES_REQUESTED** [review]
just some slight changes

**Review by sovdeeth — APPROVED** [review]

File: src/main/resources/config.sk [hunk 30]
```diff
@@ -56,6 +56,10 @@ release channel: @release-channel@
 # 'prerelease' means that development releases will also be checked for.
 # Initial value of this depends on the Skript version you first installed; it was '@release-channel@'.
 
+send reloading info to ops: false
+# Whether Skript should alert operators when someone reloads a script file.
+# This will look like the normal message you receive when reloading a script, containing reload time, errors and/or warnings.
```
Comment (sovdeeth) [inline src/main/resources/config.sk:61]: This isn't really necessary anymore since it's gated behind an opt-in permission

**Review by sovdeeth — CHANGES_REQUESTED** [review]

File: src/main/resources/config.sk (hunk 30 above)
Comment (Asleeepp) [reply src/main/resources/config.sk:61]: You want me to remove the config?

Comment (sovdeeth) [reply src/main/resources/config.sk:61]: yes it's not necessary
otherwise you have to both enable this and give a permission to a player

**Review by Moderocky — APPROVED** [review]

**Review by sovdeeth — APPROVED** [review]

**Comment by EquipableMC** [comment]
Reopening this, redirectingloghandler breaks stuff, I don't know why this was merged.

image
